File: frontend/src/app/services/salary-component.service.ts
import { Injectable } from '@angular/core';
import { HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { ApiService } from './api.service';
import { SalaryComponentRequest, SalaryComponentResponse } from '../models/salary-component.model';
import { Page } from '../models/page.model';

@Injectable({ providedIn: 'root' })
export class SalaryComponentService {
  private path = '/salary-components';

  constructor(private api: ApiService) {}

  getAll(page = 0, size = 10): Observable<Page<SalaryComponentResponse>> {
    const params = new HttpParams().set('page', page).set('size', size);
    return this.api.get<Page<SalaryComponentResponse>>(this.path, params);
  }

  getById(id: number): Observable<SalaryComponentResponse> {
    return this.api.get<SalaryComponentResponse>(`${this.path}/${id}`);
  }

  getByType(type: string): Observable<SalaryComponentResponse[]> {
    return this.api.get<SalaryComponentResponse[]>(`${this.path}/type/${type}`);
  }

  getActive(): Observable<SalaryComponentResponse[]> {
    return this.api.get<SalaryComponentResponse[]>(`${this.path}/active`);
  }

  create(component: SalaryComponentRequest): Observable<SalaryComponentResponse> {
    return this.api.post<SalaryComponentResponse>(this.path, component);
  }

  update(id: number, component: SalaryComponentRequest): Observable<SalaryComponentResponse> {
    return this.api.put<SalaryComponentResponse>(`${this.path}/${id}`, component);
  }

  updateActive(id: number, active: boolean): Observable<SalaryComponentResponse> {
    return this.api.patch<SalaryComponentResponse>(`${this.path}/${id}/active`, { active });
  }

  delete(id: number): Observable<void> {
    return this.api.delete<void>(`${this.path}/${id}`);
  }
}

File: backend/src/main/java/com/raster/hrm/salarycomponent/service/SalaryComponentService.java
package com.raster.hrm.salarycomponent.service;

import com.raster.hrm.exception.BadRequestException;
import com.raster.hrm.exception.ResourceNotFoundException;
import com.raster.hrm.salarycomponent.dto.SalaryComponentRequest;
import com.raster.hrm.salarycomponent.dto.SalaryComponentResponse;
import com.raster.hrm.salarycomponent.entity.SalaryComponent;
import com.raster.hrm.salarycomponent.entity.SalaryComponentType;
import com.raster.hrm.salarycomponent.entity.SalaryComputationType;
import com.raster.hrm.salarycomponent.repository.SalaryComponentRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@Transactional
public class SalaryComponentService {

    private static final Logger log = LoggerFactory.getLogger(SalaryComponentService.class);

    private final SalaryComponentRepository salaryComponentRepository;

    public SalaryComponentService(SalaryComponentRepository salaryComponentRepository) {
        this.salaryComponentRepository = salaryComponentRepository;
    }

    @Transactional(readOnly = true)
    public Page<SalaryComponentResponse> getAll(Pageable pageable) {
        return salaryComponentRepository.findAll(pageable)
                .map(this::mapToResponse);
    }

    @Transactional(readOnly = true)
    public SalaryComponentResponse getById(Long id) {
        var component = salaryComponentRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("SalaryComponent", "id", id));
        return mapToResponse(component);
    }

    @Transactional(readOnly = true)
    public List<SalaryComponentResponse> getByType(SalaryComponentType type) {
        return salaryComponentRepository.findByType(type).stream()
                .map(this::mapToResponse)
                .toList();
    }

    @Transactional(readOnly = true)
    public List<SalaryComponentResponse> getActive() {
        return salaryComponentRepository.findByActive(true).stream()
                .map(this::mapToResponse)
                .toList();
    }

    public SalaryComponentResponse create(SalaryComponentRequest request) {
        if (salaryComponentRepository.existsByCode(request.code())) {
            throw new BadRequestException("Salary component with code '" + request.code() + "' already exists");
        }

        var component = new SalaryComponent();
        component.setCode(request.code());
        component.setName(request.name());
        component.setType(SalaryComponentType.valueOf(request.type()));
        component.setComputationType(SalaryComputationType.valueOf(request.computationType()));
        component.setPercentageValue(request.percentageValue());
        component.setTaxable(request.taxable());
        component.setMandatory(request.mandatory());
        component.setDescription(request.description());

        var saved = salaryComponentRepository.save(component);
        log.info("Created salary component with id: {} code: {}", saved.getId(), saved.getCode());
        return mapToResponse(saved);
    }

    public SalaryComponentResponse update(Long id, SalaryComponentRequest request) {
        var component = salaryComponentRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("SalaryComponent", "id", id));

        if (!component.getCode().equals(request.code()) && salaryComponentRepository.existsByCode(request.code())) {
            throw new BadRequestException("Salary component with code '" + request.code() + "' already exists");
        }

        component.setCode(request.code());
        component.setName(request.name());
        component.setType(SalaryComponentType.valueOf(request.type()));
        component.setComputationType(SalaryComputationType.valueOf(request.computationType()));
        component.setPercentageValue(request.percentageValue());
        component.setTaxable(request.taxable());
        component.setMandatory(request.mandatory());
        component.setDescription(request.description());

        var saved = salaryComponentRepository.save(component);
        log.info("Updated salary component with id: {}", saved.getId());
        return mapToResponse(saved);
    }

    public SalaryComponentResponse updateActive(Long id, boolean active) {
        var component = salaryComponentRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("SalaryComponent", "id", id));

        component.setActive(active);
        var saved = salaryComponentRepository.save(component);
        log.info("Updated active status of salary component id: {} to {}", id, active);
        return mapToResponse(saved);
    }

    public void delete(Long id) {
        var component = salaryComponentRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("SalaryComponent", "id", id));
        salaryComponentRepository.delete(component);
        log.info("Deleted salary component with id: {}", id);
    }

    private SalaryComponentResponse mapToResponse(SalaryComponent component) {
        return new SalaryComponentResponse(
                component.getId(),
                component.getCode(),
                component.getName(),
                component.getType().name(),
                component.getComputationType().name(),
                component.getPercentageValue(),
                component.isTaxable(),
                component.isMandatory(),
                component.getDescription(),
                component.isActive(),
                component.getCreatedAt(),
                component.getUpdatedAt()
        );
    }
}
